### ISSUE TITLE
Data Inconsistency When Players Switch Servers with MySQL Sync

### ISSUE THREAD

**Issue by xiaozhangup** [VolmitSoftware/Adapt#514]
**Description**  
I encountered an issue when using MySQL to sync player data: there is a chance that data fails to synchronize properly when players switch between servers quickly, resulting in data loss. This is particularly noticeable in scenarios with frequent server switching and negatively impacts the player experience.

**Steps to Reproduce**  
1. Configure MySQL as the database for player data synchronization.  
2. Have a player switch between two servers rapidly.  
3. Observe the data in the database and on the new server; occasionally, the data may be missing or out of sync.  

**Expected Behavior**  
I would like a mechanism (e.g., a locking mechanism) to ensure data consistency during server switches, so that no data is lost in the process.

**Proposed Solution**  
Consider implementing a locking mechanism during the synchronization process to prevent related operations or reads until the player's data is fully synchronized. This would help ensure that data remains consistent when players switch servers.

**Environment Details**  
- Plugin version:  1.15.2 for 1.19.2-1.21.1
- Server software and version:  Paper-1.21.1

Please let me know if more information or debugging assistance is needed!

**Comment by illyrius666** [comment]
does anything show up in the logs? or is it only visible in the database?

**Comment by xiaozhangup** [comment]
> does anything show up in the logs? or is it only visible in the database?

Neither the database nor the server logs show anything; the data is simply lost.

**Comment by illyrius666** [comment]
> > does anything show up in the logs? or is it only visible in the database?
> 
> Neither the database nor the server logs show anything; the data is simply lost.

if you dont mind and join the discord server, please create a ticket in the support forum, so the devs can help you more directly:
https://discord.gg/volmit

**Comment by AvianAeternum** [comment]
What's providing the MySQL server

**Comment by xiaozhangup** [comment]
> What's providing the MySQL db. is it hosted on the server, or is it remote or?

on server,not remote

**Comment by CrazyDev05** [comment]
this is kinda expected as cross server sync is difficult by default and adapt wasn't even remotely designed for that

**Comment by xiaozhangup** [comment]
> this is kinda expected as cross server sync is difficult by default and adapt wasn't even remotely designed for that

I used to write plugins that required cross-server communication through a database, and I encountered the same issue where data couldn't synchronize correctly. My solution was to add a timestamp to the database to determine whether the current data was being used by any server. If it was occupied, I would wait until the occupation was released before loading the data.

**Comment by xiaozhangup** [comment]
> this is kinda expected as cross server sync is difficult by default and adapt wasn't even remotely designed for that

I just took a look at the implementation of Adapt, and it seems that using locking in this way would result in extensive code modifications. This is because it appears that the possibility of player data not being loaded yet during skill handling was not considered.🤔

**Comment by CrazyDev05** [comment]
> I just took a look at the implementation of Adapt, and it seems that using locking in this way would result in extensive code modifications. This is because it appears that the possibility of player data not being loaded yet during skill handling was not considered.🤔

yep exactly what I meant with `adapt wasn't even remotely designed for that`
if you wanna add it go ahead try but for now this will be a low prio due to it's complexity and relatively rare use case

**Comment by CrazyDev05** [comment]
if you servers are on the same machine, it's probably the easiest to just use the file storage with symbolic links
_yes i know thats very janky_

**Comment by xiaozhangup** [comment]
> if you servers are on the same machine, it's probably the easiest to just use the file storage with symbolic links _yes i know thats very janky_
> 
> but even that may have sync problems, which are at least relatively easy to fix

Actually, this would already cause the same issue as using the data, because the problem lies in the fact that the data isn't being saved in a timely manner. Could you tell me under what circumstances Adapt typically starts saving data?

**Comment by CrazyDev05** [comment]
> Actually, this would already cause the same issue as using the database, because the problem lies in the fact that the data isn't being saved in a timely manner. Could you tell me under what circumstances Adapt typically starts saving data?

the current code yes but if you use a file lock only one server will be able to own the data, making desync impossible
the are two moments adapt saves the playerdata:
1. once a minute
2. on player quit event
